src/router/open-plc-utils/pib/getpib.1
.TH getpib 1 "November 2013" "open-plc-utils-0.0.3" "Qualcomm Atheros Open Powerline Toolkit"

.SH NAME
getpib - Qualcomm Atheros PIB Data Extractor 

.SH SYNOPSIS
.BR getpib 
.IR file 
.IR offset 
.IR format 
.RI [ length ]

.SH DESCRIPTION
This program is a companion to program \fBsetpib\fR and is designed to support shell scripts.
It extracts PIB file parameters so that a scripting language can evaluate, manipulate or store them with minimal effort.
Good scripting skills and some imagination may be required to make effective use of this program.

.PP
Users having a detailed knowledge of PIB content and structure can make extremely effective use of this program; however, Atheros is under no obligation to provide customers with the offset, length or format of PIB file parameters.

.PP
This program is part of the Qualcomm Atheros Powerline Toolkit.
See the \fBAMP\fR man page for an overview and installation instructions.

.SH OPTIONS

.TP
- \fBc \fIcharacter\fR
The appended character.
Append this character each line of output instead of newline.
Use this to create a comma separated list of value, for example, by specifying a comma.

.TP
.RB - n
Append a newline to the output.

.TP
.RB - q
Enable quiet mode which has no effect.

.TP
.RB - v
Verbose mode.
Does nothing.

.TP
.RB - ? ,-- help
Print program help summary on stdout.
This option takes precedence over other options on the command line.

.TP
.RB - ! ,-- version
Print program version information on stdout.
This option takes precedence over other options on the command line.
Usethis option when sending screen dumps to Atheros Technical Support so that they know exactly which version of the Linux Toolkit you are using.

.SH ARGUMENTS

.TP
.IR file
The PIB filename.
This argument is required and must appear first on the command line.
No assumptions are made based on filename and no filename conventions are enforced; however, an invalid PIB file will be rejected.

.TP
.IR offset
The data offset in bytes expressed in hexadecimal.
This argument is required and must follow the filename.
A leading "\fB0x\fR" prefix is optional.
The data \fIoffset\fR plus the data \fIlength\fR cannot exceed the file extent.

.TP
.IB format
The output format.
This argument is required and must follow the offset.
Some formats imply fixed data length.
Others formats are for variable length data.
The fixed data formats are \fBbyte\fR, \fBword\fR, \fBlong\fR, \fBhfid\fR, \fBmac\fR and \fBkey\fR.
The variable data formats are \fBdata\fR and \fBtext\fR.
The formats are described in the next section.

.TP
.IB length
The data length in bytes expressed in decimal.
This argument is only required for the the data formats \fBdata\fR and \fBtext\fR.
The data \fIoffset\fR plus the data \fIlength\fR cannot exceed the file extent.
The data \fIlength\fR cannot exceed \fB1024\fR.

.SH FORMATS

.TP
.BR byte
Extract 1 byte and display it as an unsigned decimal integer value.
The offset advances \f1\fR byte before starting another extraction.

.TP
.BR word
Extract 2 consecutive bytes and display them as an unsigned decimal integer value after endian conversion.
The offset advances \f2\fR bytes before starting another extraction.

.TP
.BR long
Extract 4 consecutive bytes and display them as an unsigned decimal integer value after endian conversion.
The offset advances \fB4\fR bytes before starting another extraction.

.TP
.BR huge
Extract 8 consecutive bytes and display them as an unsigned decimal integer value after endian conversion.
The offset advances \fB8\fR bytes before starting another extraction.

.TP
.BR xbyte
Extract 1 byte and display it as a hexadecimal integer value.
The offset advances \f1\fR byte before starting another extraction.

.TP
.BR xword
Extract 2 consecutive bytes and display them as a hexadecimal integer value after endian conversion.
The offset advances \f2\fR bytes before starting another extraction.

.TP
.BR xlong
Extract 4 consecutive bytes and display them as a hexadecimal integer value after endian conversion.
The offset advances \fB4\fR bytes before starting another extraction.

.TP
.BR xhuge
Extract 8 consecutive bytes and display them as a hexadecimal integer value after endian conversion.
The offset advances \fB8\fR bytes before starting another extraction.

.TP
\fBdata \fIlength\fR
Extract the specified number of bytes and display them as a hexadecimal string with no punctuation to separate octets.
The offset advances \fIlength\fR bytes before starting another extraction.
The minimum length is \fB1\fR.
The maximum length is \fB1024\fR.

.TP
.BR mac
Extract 6 bytes and display them as a colon-separated hexadecimal string.
The offset advances 6 bytes before starting another extraction.
This is similar to "\fBdata 6\fR" but colons separate each octet.

.TP
.BR key
Extract 16 bytes and display them as a colon-separated hexadecimal string.
The offset advances 16 bytes before starting another extraction.
This is similar to "\fBdata 16\fR" but colons separate each octet.

.TP
.BR hfid
Extract 64 consecutive bytes and display them as an ASCII string that terminates at the first non-printable character.
The offset advances \fB64\fR bytes before starting another extraction.
This is equivalent to "\fBtext 64\fR".

.TP
\fBskip \fIlength\fR
Advance the specified number of bytes without displaying anything.
The offset advances by \fIlength\fR bytes before starting another extraction.
The minimum length is \fB1\fR.
The maximum length is \fB1024\fR.

.TP
\fBtext \fIlength\fR
Extract the specified number of bytes and display them as an ASCII string that terminates at the first non-printable character.
The offset advances the specified number of bytes before starting another extraction.

.SH TR-069

.TP
.BR accesspassword
Extract 65 consecutive bytes and display them as an ASCII string that terminates at the first non-printable character.
The offset advances \fB65\fR bytes before starting another extraction.
This format is equivalent to "\fBtext 65\fR".

.TP
.BR accessusername
Extract 33 consecutive bytes and display them as an ASCII string that terminates at the first non-printable character.
The offset advances \fB33\fR bytes before starting another extraction.
This format is equivalent to "\fBtext 33\fR".

.TP
.BR adminpassword
Extract 33 consecutive bytes and display them as an ASCII string that terminates at the first non-printable character.
The offset advances \fB33\fR bytes before starting another extraction.
This format is equivalent to "\fBtext 33\fR".

.TP
.BR adminusername
Extract 33 consecutive bytes and display them as an ASCII string that terminates at the first non-printable character.
The offset advances \fB33\fR bytes before starting another extraction.
This format is equivalent to "\fBtext 33\fR".

.TP
.BR password
Extract 257 consecutive bytes and display them as an ASCII string that terminates at the first non-printable character.
The offset advances \fB257\fR bytes before starting another extraction.
This format is equivalent to "\fBtext 257\fR".

.TP
.BR url
Extract 257 consecutive bytes and display them as an ASCII string that terminates at the first non-printable character.
The offset advances \fB257\fR bytes before starting another extraction.
This format is equivalent to "\fBtext 257\fR".

.TP
.BR username
Extract 257 consecutive bytes and display them as an ASCII string that terminates at the first non-printable character.
The offset advances \fB257\fR bytes before starting another extraction.
This format is equivalent to "\fBtext 257\fR".

.SH EXAMPLES
The following example extracts one byte from offset \fB01F7\fR of PIB file \fBabc.pib\fR and displays it as an unsigned decimal integer string.
No length specification is needed because the \fBbyte\fR format has an implied length of \fB1\fR byte.
The displayed value is \fB232\fR because the \fBbyte\fR format is decimal.
We could have specified "\fBdata 1\fR" to display the byte in hexadecimal.
The return prompt appears on the display line because option \fB-n\fR was omitted.

.PP
   # getpib abc.pib 01F7 byte 
   232#

.PP
The next example extracts two bytes at offset \fB01F7\fR and displays them as a hexadecimal string.
A length of \fB2\fR is needed because the \fBdata\fR format is variable length.
We could have specified "\fBword\fR" to display these bytes as an unsigned decimal integer string.
The hexadecimal string consists two octets \fBE8\fR and \fB8A\fR.
The first byte is the same one extracted in the last example.
The return prompt appears on a new line because option \fB-n\fR was present.

.PP
   # getpib abc.pib 01F7 data 2 -n
   E88A
   #

.SH DISCLAIMER
PIB file structure and content is proprietary to Qualcomm Atheros, Ocala FL USA.
Consequently, public information is not available.
Qualcomm Atheros reserves the right to modify PIB file structure or content in future firmware releases without any obligation to notify or compensate users of this program.

.SH SEE ALSO
.BR chkpib (7),
.BR chkpib2 (7),
.BR modpib ( 1 ),
.BR pib2xml ( 1 ),
.BR pibcomp ( 1 ),
.BR pibdump ( 1 ),
.BR setpib ( 1 ),
.BR xml2pib ( 1 )

.SH CREDITS
 Charles Maier <cmaier@qca.qualcomm.com>



src/router/open-plc-utils/pib/getpib.c
/*====================================================================*
 *
 *   Copyright (c) 2013 Qualcomm Atheros, Inc.
 *
 *   All rights reserved.
 *
 *   Redistribution and use in source and binary forms, with or 
 *   without modification, are permitted (subject to the limitations 
 *   in the disclaimer below) provided that the following conditions 
 *   are met:
 *
 *   * Redistributions of source code must retain the above copyright 
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above 
 *     copyright notice, this list of conditions and the following 
 *     disclaimer in the documentation and/or other materials 
 *     provided with the distribution.
 *
 *   * Neither the name of Qualcomm Atheros nor the names of 
 *     its contributors may be used to endorse or promote products 
 *     derived from this software without specific prior written 
 *     permission.
 *
 *   NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE 
 *   GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE 
 *   COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR 
 *   IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED 
 *   WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
 *   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER 
 *   OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
 *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT 
 *   NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
 *   LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) 
 *   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
 *   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE 
 *   OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS 
 *   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.  
 *
 *--------------------------------------------------------------------*/

/*====================================================================*
 *
 *   getpib.c - PIB Data Extractor
 *
 *   Contributor(s):
 *      Charles Maier <cmaier@qca.qualcomm.com>
 *
 *--------------------------------------------------------------------*/

/*====================================================================*
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>
#include <ctype.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/memory.h"
#include "../tools/number.h"
#include "../tools/error.h"
#include "../tools/types.h"
#include "../tools/flags.h"
#include "../tools/files.h"
#include "../pib/pib.h"
#include "../nvm/nvm.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/uintspec.c"
#include "../tools/basespec.c"
#include "../tools/todigit.c"
#include "../tools/hexout.c"
#include "../tools/error.c"
#include "../tools/checksum32.c"
#include "../tools/fdchecksum32.c"
#endif

#ifndef MAKEFILE
#include "../nvm/nvmseek2.c"
#endif

/*====================================================================*
 *   constants;
 *--------------------------------------------------------------------*/

#define GETPIB_COMMA ' '
#define GETPIB_TOOBIG "object '%s' exceeds extent of " SIZE_T_SPEC " bytes"
#define GETPIB_NOSIZE "object '%s' has no length"

#define GETPIB_VERBOSE (1 << 0)
#define GETPIB_SILENCE (1 << 1)
#define GETPIB_NEWLINE (1 << 2)

/*====================================================================*
 *
 *   void getmemory (byte const * memory, size_t extent, char const * object, size_t length);
 *
 *--------------------------------------------------------------------*/

static void getmemory (byte const * memory, size_t extent, char const * object, size_t length) 

{ 
	if (length > extent) 
	{ 
		error (1, ECANCELED, GETPIB_TOOBIG, object, length); 
	} 
	hexout (memory, length, ':', '\0', stdout); 
	return; 
} 

/*====================================================================*
 *
 *   void getstring (byte const * memory, size_t extent, char const * object, size_t length);
 *
 *--------------------------------------------------------------------*/

static void getstring (byte const * memory, size_t extent, char const * object, size_t length) 

{ 
	char const * string = (char const *) (memory); 
	if (length > extent) 
	{ 
		error (1, ECANCELED, GETPIB_TOOBIG, object, length); 
	} 
	while (isprint (* string) && (length--)) 
	{ 
		putc (* string++, stdout); 
	} 
	return; 
} 

/*====================================================================*
 *
 *   void snatch (int argc, char const * argv [], byte const * memory, size_t extent, char comma);
 *
 *   extract and print the specified data objects from memory; comma
 *   delimits consecutive objects on output; 
 *
 *--------------------------------------------------------------------*/

static void snatch (int argc, char const * argv [], byte const * memory, size_t extent, char comma) 

{ 
	size_t length = 0; 
	size_t offset = 0; 
	if (! (argc) || ! (* argv)) 
	{ 
		error (1, ECANCELED, "Need an offset"); 
	} 
	offset = (size_t) (basespec (* argv, 16, sizeof (uint32_t))); 
	if (offset > extent) 
	{ 
		error (1, ECANCELED, "offset " SIZE_T_SPEC " exceeds extent of " SIZE_T_SPEC " bytes", offset, extent); 
	} 
	memory += offset; 
	extent -= offset; 
	argc--; 
	argv++; 
	while ((argc) && (* argv)) 
	{ 
		char const * object = * argv; 
		argc--; 
		argv++; 
		if (! strcmp (object, "byte")) 
		{ 
			uint8_t * number = (uint8_t *) (memory); 
			if (sizeof (* number) > extent) 
			{ 
				error (1, ECANCELED, GETPIB_TOOBIG, object, extent); 
			} 
			printf ("%u", * number); 
			memory += sizeof (* number); 
			extent -= sizeof (* number); 
		} 
		else if (! strcmp (object, "word")) 
		{ 
			uint16_t * number = (uint16_t *) (memory); 
			if (sizeof (* number) > extent) 
			{ 
				error (1, ECANCELED, GETPIB_TOOBIG, object, extent); 
			} 
			printf ("%u", LE16TOH (* number)); 
			memory += sizeof (* number); 
			extent -= sizeof (* number); 
		} 
		else if (! strcmp (object, "long")) 
		{ 
			uint32_t * number = (uint32_t *) (memory); 
			if (sizeof (* number) > extent) 
			{ 
				error (1, ECANCELED, GETPIB_TOOBIG, object, extent); 
			} 
			printf ("%u", LE32TOH (* number)); 
			memory += sizeof (* number); 
			extent -= sizeof (* number); 
		} 
		else if (! strcmp (object, "huge")) 
		{ 
			uint64_t * number = (uint64_t *) (memory); 
			if (sizeof (* number) > extent) 
			{ 
				error (1, ECANCELED, GETPIB_TOOBIG, object, extent); 
			} 
			printf ("%llu", LE64TOH (* number)); 
			memory += sizeof (* number); 
			extent -= sizeof (* number); 
		} 

#if 1

		else if (! strcmp (object, "xbyte")) 
		{ 
			uint8_t * number = (uint8_t *) (memory); 
			if (sizeof (* number) > extent) 
			{ 
				error (1, ECANCELED, GETPIB_TOOBIG, object, extent); 
			} 
			printf ("0x%02X", * number); 
			memory += sizeof (* number); 
			extent -= sizeof (* number); 
		} 
		else if (! strcmp (object, "xword")) 
		{ 
			uint16_t * number = (uint16_t *) (memory); 
			if (sizeof (* number) > extent) 
			{ 
				error (1, ECANCELED, GETPIB_TOOBIG, object, extent); 
			} 
			printf ("0x%04X", LE16TOH (* number)); 
			memory += sizeof (* number); 
			extent -= sizeof (* number); 
		} 
		else if (! strcmp (object, "xlong")) 
		{ 
			uint32_t * number = (uint32_t *) (memory); 
			if (sizeof (* number) > extent) 
			{ 
				error (1, ECANCELED, GETPIB_TOOBIG, object, extent); 
			} 
			printf ("0x%08X", LE32TOH (* number)); 
			memory += sizeof (* number); 
			extent -= sizeof (* number); 
		} 
		else if (! strcmp (object, "xhuge")) 
		{ 
			uint64_t * number = (uint64_t *) (memory); 
			if (sizeof (* number) > extent) 
			{ 
				error (1, ECANCELED, GETPIB_TOOBIG, object, extent); 
			} 
			printf ("0x%016llX", LE64TOH (* number)); 
			memory += sizeof (* number); 
			extent -= sizeof (* number); 
		} 

#endif

		else if (! strcmp (object, "mac")) 
		{ 
			length = ETHER_ADDR_LEN; 
			if (length > extent) 
			{ 
				error (1, ECANCELED, GETPIB_TOOBIG, object, extent); 
			} 
			getmemory (memory, extent, object, length); 
			memory += length; 
			extent -= length; 
		} 
		else if (! strcmp (object, "key")) 
		{ 
			length = PIB_KEY_LEN; 
			if (length > extent) 
			{ 
				error (1, ECANCELED, GETPIB_TOOBIG, object, extent); 
			} 
			getmemory (memory, extent, object, length); 
			memory += length; 
			extent -= length; 
		} 
		else if (! strcmp (object, "hfid")) 
		{ 
			length = PIB_HFID_LEN; 
			if (length > extent) 
			{ 
				error (1, ECANCELED, GETPIB_TOOBIG, object, extent); 
			} 
			getstring (memory, extent, object, length); 
			memory += length; 
			extent -= length; 
		} 

#if 1

		else if (! strcmp (object, "adminusername") || ! strcmp (object, "adminpassword") || ! strcmp (object, "accessusername")) 
		{ 
			length = PIB_NAME_LEN +  1; 
			if (length > extent) 
			{ 
				error (1, ECANCELED, GETPIB_TOOBIG, object, extent); 
			} 
			getstring (memory, extent, object, length); 
			memory += length; 
			extent -= length; 
		} 
		else if (! strcmp (object, "accesspassword")) 
		{ 
			length = PIB_HFID_LEN +  1; 
			if (length > extent) 
			{ 
				error (1, ECANCELED, GETPIB_TOOBIG, object, extent); 
			} 
			getstring (memory, extent, object, length); 
			memory += length; 
			extent -= length; 
		} 
		else if (! strcmp (object, "username") || ! strcmp (object, "password") || ! strcmp (object, "url")) 
		{ 
			length = PIB_TEXT_LEN +  1; 
			getstring (memory, extent, object, length); 
			memory += length; 
			extent -= length; 
		} 

#endif

		else if (! strcmp (object, "data")) 
		{ 
			if (! * argv) 
			{ 
				error (1, EINVAL, GETPIB_NOSIZE, object); 
			} 
			length = (unsigned) (uintspec (* argv, 1, extent)); 
			hexout (memory, length, 0, 0, stdout); 
			memory += length; 
			extent -= length; 
			argc--; 
			argv++; 
		} 
		else if (! strcmp (object, "text")) 
		{ 
			if (! * argv) 
			{ 
				error (1, EINVAL, GETPIB_NOSIZE, object); 
			} 
			length = (unsigned) (uintspec (* argv, 1, extent)); 
			getstring (memory, extent, object, length); 
			memory += length; 
			extent -= length; 
			argc--; 
			argv++; 
		} 
		else if (! strcmp (object, "skip")) 
		{ 
			if (! * argv) 
			{ 
				error (1, EINVAL, GETPIB_NOSIZE, object); 
			} 
			length = (unsigned) (uintspec (* argv, 1, extent)); 
			memory += length; 
			extent -= length; 
			argc--; 
			argv++; 
			continue; 
		} 
		else 
		{ 
			error (1, ENOTSUP, "%s", object); 
		} 
		if ((argc) && (* argv)) 
		{ 
			putc (comma, stdout); 
		} 
	} 
	return; 
} 

/*====================================================================*
 *
 *   signed pibimage1 (int argc, char const * argv [], char comma);
 * 
 *   read an entire flat parameter file into memory, edit it, save 
 *   it and display it;
 *
 *   Contributor(s):
 *	Charles Maier <cmaier@qca.qualcomm.com>
 *
 *--------------------------------------------------------------------*/

static signed pibimage1 (int argc, char const * argv [], char comma) 

{ 
	signed fd; 
	off_t extent; 
	byte * memory; 
	if ((fd = open (* argv, O_BINARY | O_RDWR)) == - 1) 
	{ 
		error (1, errno, FILE_CANTOPEN, * argv); 
	} 
	if ((extent = lseek (fd, 0, SEEK_END)) == - 1) 
	{ 
		error (1, errno, FILE_CANTSIZE, * argv); 
	} 
	if (lseek (fd, 0, SEEK_SET)) 
	{ 
		error (1, errno, FILE_CANTHOME, * argv); 
	} 
	if (! (memory = malloc (extent))) 
	{ 
		error (1, errno, FILE_CANTLOAD, * argv); 
	} 
	if (read (fd, memory, extent) != extent) 
	{ 
		error (1, errno, FILE_CANTREAD, * argv); 
	} 
	close (fd); 
	snatch (argc - 1, argv +  1, memory, extent, comma); 
	free (memory); 
	return (0); 
} 

/*====================================================================*
 *
 *   signed pibimage2 (int argc, char const * argv [], char comma);
 * 
 *   read an entire flat parameter file into memory, edit it, save 
 *   it and display it;
 *
 *   Contributor(s):
 *	Charles Maier <cmaier@qca.qualcomm.com>
 *
 *--------------------------------------------------------------------*/

static signed pibimage2 (int argc, char const * argv [], char comma) 

{ 
	struct nvm_header2 header; 
	signed fd; 
	off_t extent; 
	byte * memory; 
	if ((fd = open (* argv, O_BINARY | O_RDWR)) == - 1) 
	{ 
		error (1, errno, FILE_CANTOPEN, * argv); 
	} 
	if (nvmseek2 (fd, * argv, & header, NVM_IMAGE_PIB)) 
	{ 
		error (1, errno, "Can't find PIB image in %s", * argv); 
	} 
	extent = LE32TOH (header.ImageLength); 
	if (! (memory = malloc (extent))) 
	{ 
		error (1, errno, FILE_CANTLOAD, * argv); 
	} 
	if (read (fd, memory, extent) != extent) 
	{ 
		error (1, errno, FILE_CANTREAD, * argv); 
	} 
	close (fd); 
	snatch (argc - 1, argv +  1, memory, extent, comma); 
	free (memory); 
	return (0); 
} 

/*====================================================================*
 *
 *   signed function (int argc, char const * argv [], char comma);
 *
 *   call an appropriate parameter edit function based on the file 
 *   header;
 *
 *   older parameter files are flat with their own header; newer ones
 *   are image chains where one of image contains the parameter block;
 *
 *   
 *   Contributor(s):
 *	Charles Maier <cmaier@qca.qualcomm.com>
 *
 *--------------------------------------------------------------------*/

static signed function (int argc, char const * argv [], char comma) 

{ 
	uint32_t version; 
	signed status; 
	signed fd; 
	if ((fd = open (* argv, O_BINARY | O_RDWR)) == - 1) 
	{ 
		error (1, errno, FILE_CANTOPEN, * argv); 
	} 
	if (read (fd, & version, sizeof (version)) != sizeof (version)) 
	{ 
		error (1, errno, FILE_CANTREAD, * argv); 
	} 
	close (fd); 
	if (LE32TOH (version) == 0x00010001) 
	{ 
		status = pibimage2 (argc, argv, comma); 
	} 
	else 
	{ 
		status = pibimage1 (argc, argv, comma); 
	} 
	return (status); 
} 

/*====================================================================*
 *
 *   int main (int argc, char const * argv []);
 *   
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv []) 

{ 
	static char const * optv [] = 
	{ 
		"c:qvn", 
		"file offset type [size]\n\n\tstandard-length types are 'byte'|'word'|'long'|'huge'|'hfid'|'mac'|'key'\n\tvariable-length types are 'data'|'text'|'skip' and need a size", 
		"PIB Data Extractor", 
		"c c\tobject separator is (c) [" LITERAL (GETPIB_COMMA) "]", 
		"n\tappend newline", 
		"q\tquiet mode", 
		"v\tverbose mode", 
		(char const *) (0)
	}; 
	flag_t flags = (flag_t) (0); 
	char comma = GETPIB_COMMA; 
	signed c; 
	optind = 1; 
	opterr = 1; 
	while (~ (c = getoptv (argc, argv, optv))) 
	{ 
		switch (c) 
		{ 
		case 'c': 
			comma = * optarg; 
			break; 
		case 'n': 
			_setbits (flags, GETPIB_NEWLINE); 
			break; 
		case 'q': 
			_setbits (flags, GETPIB_SILENCE); 
			break; 
		case 'v': 
			_setbits (flags, GETPIB_VERBOSE); 
			break; 
		default: 
			break; 
		} 
	} 
	argc -= optind; 
	argv += optind; 
	if ((argc) && (* argv)) 
	{ 
		function (argc, argv, comma); 
		if (_anyset (flags, GETPIB_NEWLINE)) 
		{ 
			putc ('\n', stdout); 
		} 
	} 
	return (0); 
} 
